File: java/problems/p0020_valid_parentheses/Solution.java
package problems.p0020_valid_parentheses;

class Solution {
    public boolean isValid(String s) {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < s.length(); i++) {
            char current = s.charAt(i);
            if (current == '(' || current == '[' || current == '{') {
                sb.append(current);
            } else {
                if (sb.length() == 0) {
                    return false;
                }

                char last = sb.charAt(sb.length() - 1);
                if ((current == ')' && last == '(') ||
                    (current == ']' && last == '[') ||
                    (current == '}' && last == '{')) {
                        sb.deleteCharAt(sb.length() - 1);
                } else {
                    return false;
                }
            }
        }

        return sb.length() == 0;
    }
}

File: typescript/problems/p0020_valid_parentheses/solution.ts
function isValid(s: string): boolean {
  let stack: string[] = [];

  for (let i = 0; i < s.length; i++) {
    let current: string = s[i];

    if (current === "(" || current === "[" || current === "{") {
      stack.push(current);
    } else {
      if (stack.length === 0) return false;

      let last: string = stack[stack.length - 1];
      if (
        (current === ")" && last === "(") ||
        (current === "]" && last === "[") ||
        (current === "}" && last === "{")
      ) {
        stack.pop();
      } else {
        return false;
      }
    }
  }

  return stack.length === 0;
}

export default isValid;
